### PR TITLE
snopt c++ interface for the optimization problem class

### PR DESCRIPTION
**Pull request by RussTedrake**
implements the first pieces of the function class hierarchy
moved the getInitialState logic over to optimization problem (from rigid body IK)
resolves the infectious dependency on snopt

### CODE REVIEW

File: drake/core/Function.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      CONSTANT,
+      ZERO
+      // todo: add more (rational, ...)
+    };
+    Form form;
+    // todo: add sparsity info
+
+    InputOutputRelation(Form f) : form(f) { };
+
+    static bool isa(const Form &f, const Form &base) {
```
Comment (tkoolen) [inline drake/core/Function.h]: isA? Sticking to the code formatting rules is more important than using the same name as Matlab in my opinion.

File: drake/core/Function.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      Form ret = *iter;
+      while (++iter != forms.end()) {
+        ret = leastCommonAncestor(ret, *iter);
+      }
+      return ret;
+    }
+
+    static InputOutputRelation composeWith(const InputOutputRelation &g,
+                                           const InputOutputRelation &f) { // composition of functions y = g(f(x))
+      return InputOutputRelation(leastCommonAncestor(g.form, f.form));
```
Comment (tkoolen) [inline drake/core/Function.h]: Don't know if we're going to care about this case, but imagine composing an arbitrary function and a zero function. The best type of the result would be a zero function, not an arbitrary function.

File: drake/core/Gradient.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  typedef TaylorVard<Eigen::Dynamic> TaylorVarXd;
+  typedef TaylorVecd<Eigen::Dynamic,Eigen::Dynamic> TaylorVecXd;
+  typedef TaylorMatd<Eigen::Dynamic,Eigen::Dynamic,Eigen::Dynamic> TaylorMatXd;
+
+  /** initTaylorVecXd
+   * @brief initializes the vector with x=val and dx=eye(numel(val))
+   */
+  template <typename DerivedA, typename DerivedB>
+  void initTaylorVecXd(const Eigen::MatrixBase<DerivedA>& val, Eigen::MatrixBase<DerivedB>& auto_diff_vec) {
+    Eigen::MatrixXd der = Eigen::MatrixXd::Identity(val.rows(),val.rows());
```
Comment (tkoolen) [inline drake/core/Gradient.h]: Creating this matrix is a bit of an expensive way to do it.
This AutoDiffScalar constructor might be handy:
```
    /** Constructs an active scalar from its \a value,
        and initializes the \a nbDer derivatives such that it corresponds to the \a derNumber -th variable */
    AutoDiffScalar(const Scalar& value, int nbDer, int derNumber)
      : m_value(value), m_derivatives(DerType::Zero(nbDer))
    {
      m_derivatives.coeffRef(derNumber) = Scalar(1);
    }
```

Comment (tkoolen) [inline drake/core/Gradient.h]: replacing rows() by size() throughout this function would make it work for arbitrarily sized matrices, not just vectors.

Comment (tkoolen) [inline drake/core/Gradient.h]: I think it would also be better to infer the scalar type of the derivatives from auto_diff_vec instead of assuming it's double (by creating a MatrixXd). That would make the function naturally support higher order derivatives.

File: drake/core/Gradient.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    initTaylorVecXd(val, auto_diff_vec);
+    return auto_diff_vec;
+  }
+
+
+  /** initTaylorVecXd
+   * @brief initializes the vector with x=value and dx=gradient
+   */
+  template<typename Derived, typename DerivedGradient>
+  void initTaylorVecXd(const Eigen::MatrixBase<Derived>& val, const Eigen::MatrixBase<DerivedGradient>& gradient, TaylorVecXd& auto_diff_vector)
```
Comment (tkoolen) [inline drake/core/Gradient.h]: I've been doing this using .cast<bla>(), followed by a call to gradientMatrixToAutoDiff defined in drakeGradientUtil. I like this better and would like to change stuff to use this, but this doesn't support arbitrarily sized matrices, only vectors. Again it's trivial to support arbitrarily sized matrices by changing .rows() to .size().

File: drake/core/Vector.h [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+  /** NullVector<ScalarType>
+   * @brief provides the empty vector (templated by ScalarType)
+   * @concept{vector_concept}
+   */
+  template<typename ScalarType> using NullVector = Eigen::Matrix<ScalarType, 0, 1>;
+
+
+  template<typename ScalarType, int Rows>
+  Eigen::Matrix<ScalarType, Rows, 1> toEigen(const Eigen::Matrix<ScalarType, Rows, 1> &vec) { return vec; }
```
Comment (tkoolen) [inline drake/core/Vector.h]: Would be nice if we could return a reference from this to avoid the copy. Is that safe?

File: drake/core/Vector.h [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+  class CombinedVector {
+  public:
+    CombinedVector() { };  // allow use of default constructors for vec1 and vec2, also
+    CombinedVector(const Vector1<ScalarType> &first, const Vector2<ScalarType> &second) : vec1(first), vec2(second) { };
+
+    template<typename Derived>
+    CombinedVector(const Eigen::MatrixBase <Derived> &x) : vec1(x.topRows(Vector1<ScalarType>::RowsAtCompileTime)),
+                                                           vec2(x.bottomRows(Vector2<ScalarType>::RowsAtCompileTime)) {
+      static_assert(RowsAtCompileTime != Eigen::Dynamic,
+                    "Cannot determine sizes of subvectors because sizes are not known at compile time."); // TODO: could handle cases where only one of the subvectors has dynamic size
```
Comment (tkoolen) [inline drake/core/Vector.h]: Sorry for nagging, but I still think this is going to be a big issue.

File: drake/core/Vector.h [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+    template<typename ScalarType>
+    static NullVector <ScalarType> first(const Vector2<ScalarType> &vec) { return NullVector<ScalarType>(); }
+
+    template<typename ScalarType>
+    static const Vector2<ScalarType> &second(const Vector2<ScalarType> &vec) { return vec; }
+
+    template<typename ScalarType>
+    static Vector2<ScalarType> combine(const Vector1<ScalarType> &vec1, const Vector2<ScalarType> &vec2) { return vec2; };
+  };
```
Comment (tkoolen) [inline drake/core/Vector.h]: We can talk about this later, but I think the CombinedVectorUtil / CombinedVectorHelper stuff is too complex; I don't understand it and was running into issues with it when trying to do a dynamic sized system. I'll open an issue on that when I get back to the Python-wrappable dynamic size system.

File: drake/examples/Cars/CMakeLists.txt [hunk 8]
```diff
@@ -7,7 +7,7 @@ include_directories(${PROJECT_SOURCE_DIR}/systems/plants/joints)
 include_directories(${PROJECT_SOURCE_DIR}/systems/controllers)
 include_directories(${PROJECT_SOURCE_DIR}/util)
 
-if (LCM_FOUND AND snopt_c_FOUND)
+if (LCM_FOUND)
```
Comment (tkoolen) [inline drake/examples/Cars/CMakeLists.txt]: Awesome.

File: drake/solvers/CMakeLists.txt [hunk 9]
```diff
@@ -32,28 +32,23 @@ if(MATLAB_FOUND)
   target_link_libraries(lqrmex drakeUtil)
 endif()
 
+set(optimization_files)
+if (snopt_c_FOUND)
+  list(APPEND optimization_files SNOPTWrapper.cpp)
+else()
+  list(APPEND optimization_files NoSNOPT.cpp)
+endif(snopt_c_FOUND)
+include_directories(${PROJECT_SOURCE_DIR}/core)
```
Comment (tkoolen) [inline drake/solvers/CMakeLists.txt]: Shouldn't do this anymore! Going to mess things up for projects that only use the Drake installed headers.

File: drake/solvers/CMakeLists.txt [hunk 10]
```diff
@@ -32,28 +32,23 @@ if(MATLAB_FOUND)
...
+  pods_use_pkg_config_packages(drakeOptimization snopt_c)
+endif(snopt_c_FOUND)
+
 if(snopt_c_FOUND AND MATLAB_FOUND)
   add_mex(NonlinearProgramSnoptmex NonlinearProgramSnoptmex.cpp)
   pods_use_pkg_config_packages(NonlinearProgramSnoptmex snopt_c)
   set_target_properties(NonlinearProgramSnoptmex PROPERTIES COMPILE_FLAGS -fPIC)
-
-  # link against matlab's copy of the gfortran library, if possible
-  find_library(GFORTRAN_LIBRARY NAMES gfortran libgfortran
```
Comment (tkoolen) [inline drake/solvers/CMakeLists.txt]: Why is this not necessary anymore?

File: drake/solvers/NoSNOPT.cpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+
+#include "Optimization.h"
+
+bool Drake::OptimizationProblem::NonlinearProgram::hasSNOPT() const { return false; }
+
+bool Drake::OptimizationProblem::NonlinearProgram::solveWithSNOPT(OptimizationProblem &prog) const {
+  throw std::runtime_error("The SNOPT bindings were not compiled.  You'll need to use a different solver.");
+  return false;
```
Comment (tkoolen) [inline drake/solvers/NoSNOPT.cpp]: Unreachable return statement.

File: drake/solvers/Optimization.h [hunk 12]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
    */
   class Constraint {
   public:
-    Constraint(const VariableList& vars) : variable_list(vars) {}
+    Constraint(const VariableList& vars, size_t num_constraints) : variable_list(vars), lower_bound(num_constraints), upper_bound(num_constraints) {
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Infer num_constraints from vars?

File: drake/solvers/Optimization.h [hunk 13]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+      assert(lower_bound.size() == upper_bound.size() && "Lower bound and upper bound must be the same size");
+      lower_bound.setConstant(-std::numeric_limits<double>::infinity());
+      upper_bound.setConstant(std::numeric_limits<double>::infinity());
+    }
+
+    template <typename DerivedLB, typename DerivedUB>
+    Constraint(const VariableList& vars, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : variable_list(vars), lower_bound(lb), upper_bound(ub)
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Check that sizes match.

File: drake/solvers/Optimization.h [hunk 14]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+    {}
     virtual ~Constraint() {}
 
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const = 0;  // move this to DifferentiableConstraint derived class if/when we need to support non-differentiable functions
+
+    virtual Eigen::VectorXd getLowerBound() const { return lower_bound; }
+    virtual Eigen::VectorXd getUpperBound() const { return upper_bound; }
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Can these be turned into const references? Going to be copying a whole lot of dynamically allocated vectors otherwise.

File: drake/solvers/Optimization.h [hunk 15]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+   */
+  class LinearConstraint : public Constraint {
+  public:
+    LinearConstraint(const VariableList& vars, size_t num_constraints) : Constraint(vars, num_constraints) {}
+    template <typename DerivedLB, typename DerivedUB>
+    LinearConstraint(const VariableList& vars, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : Constraint(vars,lb,ub) {};
+    virtual ~LinearConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=getMatrix()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = getMatrix().cast<TaylorVarXd>()*x; };
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: This doesn't look right. The derivatives won't be initialized after the cast.

File: drake/solvers/Optimization.h [hunk 16]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+    }
+
+  private:
+    const std::shared_ptr<DifferentiableFunction> func;
   };
 
-  // DifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  // todo: consider implementing DifferentiableConstraint, TwiceDifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  /** LinearConstraint
+   * @brief Implements a constraint of the form @f -lb <= Ax <= ub @f
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: -lb?

File: drake/solvers/Optimization.h [hunk 17]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+    virtual ~LinearConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=getMatrix()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = getMatrix().cast<TaylorVarXd>()*x; };
+
+    virtual Eigen::SparseMatrix<double> getSparseMatrix() const { return getMatrix().sparseView(); };
+    virtual Eigen::MatrixXd getMatrix() const = 0;
+  };
+
+  class LinearConstraintImpl : public LinearConstraint {
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Why is this indirection necessary?

File: drake/solvers/Optimization.h [hunk 18]
```diff
@@ -160,19 +245,38 @@ namespace Drake {
       if (Aeq.cols() != A.cols()) throw std::runtime_error("Can't change the number of decision variables");
       A.conservativeResize(Aeq.rows(),Eigen::NoChange);
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Why do this conservatively if we're just going to overwrite it on the next line?

File: drake/solvers/Optimization.h [hunk 19]
```diff
@@ -160,19 +245,38 @@ namespace Drake {
...
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y=x; }
 
-  class OptimizationProblem {
+    virtual Eigen::MatrixXd getMatrix() const override { return Eigen::MatrixXd::Identity(getNumConstraints(),getNumConstraints()); }
+  };
+
+  class DRAKEOPTIMIZATION_EXPORT OptimizationProblem {
   public:
-    OptimizationProblem() : problem_type(new LeastSquares), num_vars(0) {};
+    OptimizationProblem() : problem_type(new LeastSquares), num_vars(0), x_initial_guess(100) {};
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: name the magic number

File: drake/solvers/Optimization.h [hunk 20]
```diff
@@ -243,7 +430,7 @@ namespace Drake {
 //    getExitFlag();
 //    getInfeasibleConstraintNames();
 
-    void setDecisionVariableValues() {  // todo: overload the ostream operator instead?
+    void printSolution() {  // todo: overload the ostream operator instead?
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: about the todo: I like it better the way it is.

File: drake/solvers/Optimization.h [hunk 21]
```diff
@@ -260,13 +447,35 @@ namespace Drake {
       }
     }
 
+    const std::list<std::shared_ptr<Constraint>>& getGenericObjectives() const { return generic_objectives; } // e.g. for snopt_user_fun
+    const std::list<std::shared_ptr<Constraint>>& getGenericConstraints() const { return generic_constraints; } // e.g. for snopt_user_fun
+    std::list<std::shared_ptr<LinearConstraint>> getAllLinearConstraints() const {
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: So does getGenericConstraints *not* return all generic constraints?

File: drake/solvers/Optimization.h [hunk 22]
```diff
@@ -260,13 +447,35 @@ namespace Drake {
...
+      return conlist;
+    }
+
   private:
     // note: use std::list instead of std::vector because realloc in std::vector invalidates existing references to the elements
     std::list<DecisionVariable> variables;
+    VariableList variable_views;
+    std::list<std::shared_ptr<Constraint>> generic_objectives;
+    std::list<std::shared_ptr<Constraint>> generic_constraints;
+    std::list<std::shared_ptr<LinearConstraint>> linear_constraints;  // note: does not include linear_equality_constraints
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: so rename to linear_inequality_constraints?

File: drake/solvers/test/testOptimizationProblem.cpp [hunk 23]
```diff
@@ -17,21 +16,114 @@ int main(int argc, char* argv[])
...
+void sixHumpCamel() {
+  OptimizationProblem prog;
+  auto x = prog.addContinuousVariables(2);
+  auto objective = prog.addCost(make_shared<SixHumpCamelObjective>());
   prog.solve();
-  valuecheckMatrix(b.topRows(2)/2,y.value(),1e-10);
-  valuecheckMatrix(b/3,x.value(),1e-10);
+  prog.printSolution();
+
+  // check (numerically) if it is a local minima
```
Comment (tkoolen) [inline drake/solvers/test/testOptimizationProblem.cpp]: minimum

File: drake/solvers/test/testOptimizationProblem.cpp [hunk 24]
```diff
@@ -17,21 +16,114 @@ int main(int argc, char* argv[])
...
+  VectorXd ystar, y;
+  objective->eval(x.value(),ystar);
+  for (int i=0; i<10; i++) {
+    objective->eval(x.value() + .01 * Eigen::Matrix<double, 2, 1>::Random(), y);
+    if (y(0)<ystar(0)) throw std::runtime_error("not a local minima!");
+  }
+}
+
+class GPCMObjective : public TemplatedDifferentiableFunction<GPCMObjective> {
```
Comment (tkoolen) [inline drake/solvers/test/testOptimizationProblem.cpp]: What does this stand for? Google says "Grandparents.com Inc".

File: drake/solvers/test/testOptimizationProblem.cpp [hunk 25]
```diff
@@ -17,21 +16,114 @@ int main(int argc, char* argv[])
...
+  }
+};
+
+/** gloptiPolyConstrainedMinimization
+ * @brief from section 5.8.2 of the gloptipoly3 documentation
+ *
+ * Which is from section 3.5 in
+ *   Handbook of Test Problems in Local and Global Optimization
+ */
+void gloptipolyConstrainedMinimization() {
```
Comment (tkoolen) [inline drake/solvers/test/testOptimizationProblem.cpp]: aha. Still vote to rename GPCMObjective and GPCMConstraint.

File: drake/examples/Cars/simulateLCM.cpp [hunk 26]
```diff
@@ -127,6 +127,8 @@ int main(int argc, char* argv[]) {
 
   VectorXd x0(rigid_body_sys->getNumStates());
   x0.head(tree->num_positions) = tree->getZeroConfiguration();
+  // todo:  call getInitialState instead?  (but currently, that would require snopt).  needs #1627
+  // I'm getting away without it, but might be generating large internal forces initially as the ackerman constraint (hopefully) gets enforced by the stabilization terms.
 
   runLCM(sys,lcm,0,std::numeric_limits<double>::max(),x0,options);
```
Comment (tkoolen) [inline drake/examples/Cars/simulateLCM.cpp]: use infinity here too?

**Comment by tkoolen** [comment]
Done with my review. Great to get rid of snopt as a required dependency!

File: drake/core/Function.h [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+  public:
+    Function() : relation(InputOutputRelation::ARBITRARY) {};
+    Function(InputOutputRelation::Form f) : relation(f) {};
+    virtual ~Function() {};
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
+    virtual const InputOutputRelation& getInputOutputRelation() const { return relation; };
+
+  protected:
+    InputOutputRelation relation;
```
Comment (hongkai-dai) [inline drake/core/Function.h]: I thought we want to define the relation as a matrix, from input(i) to output(j)

File: drake/solvers/Optimization.h [hunk 28]
```diff
@@ -180,30 +284,89 @@ namespace Drake {
...
+      x_initial_guess.conservativeResize(num_vars);
+      x_initial_guess.tail(num_vars) = 0.1*Eigen::VectorXd::Random(num_vars);
 
-      return DecisionVariableView(variables.back());
+      return variable_views.back();
     }
 //    const DecisionVariable& addIntegerVariables(size_t num_new_vars, std::string name);
 //  ...
 
+    void addCost(const std::shared_ptr<Constraint>& obj) {
```
Comment (hongkai-dai) [inline drake/solvers/Optimization.h]: Can we add a `Function` object as the cost instead? As `Constraint` is just `Function` with lower and upper bounds, and cost does not have bounds

File: drake/solvers/SNOPTWrapper.cpp [hunk 29]
```diff
@@ -0,0 +1,343 @@
+
```
Comment (hongkai-dai) [inline drake/solvers/SNOPTWrapper.cpp]: This might be a good time to try the new SNOPT version. The interface of SNOPT is changed.

File: drake/solvers/test/testOptimizationProblem.cpp (hunk 25 above)
Comment (hongkai-dai) [inline drake/solvers/test/testOptimizationProblem.cpp]: I mistook GPCM for General Purpose Constraint Optimization in the first place.

**Comment by hongkai-dai** [comment]
Done with my code review. +1 for the code.

**Comment by RussTedrake** [comment]
thanks for the great code reviews.  will work through them today.

File: drake/core/Gradient.h (hunk 3 above)
Comment (RussTedrake) [inline drake/core/Gradient.h]: done.

File: drake/core/Gradient.h (hunk 4 above)
Comment (RussTedrake) [inline drake/core/Gradient.h]: i'm not sure exactly what the recommendation is here.  the .cast() call doesn't save much given this function signature... (would involve a copy, no?)

File: drake/core/Vector.h (hunk 6 above)
Comment (RussTedrake) [inline drake/core/Vector.h]: it's something to handle, but i don't think it blocks this PR.

File: drake/solvers/CMakeLists.txt (hunk 9 above)
Comment (RussTedrake) [inline drake/solvers/CMakeLists.txt]: good catch.  (did it before your fix)

File: drake/solvers/CMakeLists.txt (hunk 10 above)
Comment (RussTedrake) [inline drake/solvers/CMakeLists.txt]: we've been using the c only version of snopt for some time now.  compiling snopt doesn't even require fortran to be installed.

File: drake/solvers/Optimization.h (hunk 13 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: done

File: drake/solvers/Optimization.h (hunk 14 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: no.  i did that in my first implementation, but had to switch to this to support a derived class.

File: drake/core/Function.h [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+
+  /** InputOutputRelation
+   * \brief Tags which can be used to inform algorithms about underlying structure in a function
+   * e.g., linear, affine, polynomial, analytic, differentiable, continuous, measurable, and -- lastly -- arbitrary
+   *
+   * note: i considered using inheritance to capture the relationship, but passing around types at runtime
+   * was more of a pain than simply capturing the inheritance with the helper functions below.
+   */
+  struct InputOutputRelation {
+    enum Form {
```
Comment (psiorx) [inline drake/core/Function.h]: Why not use the C++11 `enum class`?

File: drake/solvers/Optimization.h [hunk 31]
```diff
@@ -57,6 +59,7 @@ namespace Drake {
      * @brief returns the actual stored value; which is only meaningful after calling solve() in the program.
      */
     Eigen::VectorBlock<const Eigen::VectorXd,-1> value() const { return var.data.segment(start_index,length); }
```
Comment (psiorx) [inline drake/solvers/Optimization.h]: Not part of this PR but shouldn't we use `Eigen::Dynamic` instead of -1?

File: drake/solvers/Optimization.h [hunk 32]
```diff
@@ -108,46 +116,123 @@ namespace Drake {
...
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=.5*x.transpose()*Q*x + b.transpose()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = .5*x.transpose()*Q.cast<TaylorVarXd>()*x + b.cast<TaylorVarXd>().transpose()*x; };
+
+  private:
+    Eigen::MatrixXd Q;
+    Eigen::VectorXd b;
+  };
+
+  class FunctionConstraint : public Constraint {
```
Comment (avalenzu) [inline drake/solvers/Optimization.h]: Should this be `DifferentiableFunctionConstraint`?

**Comment by avalenzu** [comment]
I know I'm a bit out of the loop here, so bear with me. What is the rationale for putting so many classes together in header files like `Optimization.h`?

File: drake/core/Function.h (hunk 1 above)
Comment (RussTedrake) [inline drake/core/Function.h]: done

File: drake/core/Vector.h (hunk 5 above)
Comment (RussTedrake) [inline drake/core/Vector.h]: done

File: drake/solvers/NoSNOPT.cpp (hunk 11 above)
Comment (RussTedrake) [inline drake/solvers/NoSNOPT.cpp]: zapped.

File: drake/solvers/Optimization.h (hunk 14 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: actually, i revisited this and could make things const & at least for now.  but it does add a strong  constraint on the derived classes.

File: drake/solvers/Optimization.h (hunk 15 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: I was worried about that, but thought I saw that is what you were doing in your code..?  This one is an easy fix.

File: drake/solvers/Optimization.h (hunk 16 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: good catch

File: drake/solvers/Optimization.h (hunk 17 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: To support people (soon to happen in the dynamics class) implementing a linear constraint that evaluated via computation (e.g. of H,C,B) instead of writing to A.  Perhaps I will zap it for now, and wait til I make sure we need it.

File: drake/solvers/Optimization.h (hunk 18 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: good catch.

File: drake/solvers/Optimization.h (hunk 19 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: done

File: drake/solvers/Optimization.h (hunk 21 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: I do not actually want those top two methods to exist.  What it feels like I want is to name the snopt userfun method in the cpp file as a friend.  But the problem is that it's input/output arguments depend on snopt types.

File: drake/solvers/Optimization.h (hunk 22 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: that's not *quite* right either.  i could have an object of type LinearConstraint which has lb==ub.  But it is not implemented as a derived class LinearEqualityConstraint.

File: drake/solvers/test/testOptimizationProblem.cpp (hunk 24 above)
Comment (RussTedrake) [inline drake/solvers/test/testOptimizationProblem.cpp]: that's exactly right.  alternatively, you can think of it as "GloptipolyConstrainedMinimization"; i've done the refactor to distance myself from the grandparents.

File: drake/examples/Cars/simulateLCM.cpp (hunk 26 above)
Comment (RussTedrake) [inline drake/examples/Cars/simulateLCM.cpp]: done

File: drake/solvers/Optimization.h (hunk 15 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: @tkoolen - in fact, you recommended to me to do this here:  https://codetree.com/projects/gJB8/issues?q=bug+eigen .  can we decide if the cast does the zero initialization or not?

File: drake/core/Function.h (hunk 27 above)
Comment (RussTedrake) [inline drake/core/Function.h]: added a comment to what I think the implementation might look like for that case.

File: drake/solvers/SNOPTWrapper.cpp (hunk 29 above)
Comment (RussTedrake) [inline drake/solvers/SNOPTWrapper.cpp]: can you open an issue with a few specific pointers, so that we can ask for help in trying that?

File: drake/core/Function.h (hunk 30 above)
Comment (RussTedrake) [inline drake/core/Function.h]: nice.  only because i didn't know about it.  updated.

File: drake/solvers/Optimization.h (hunk 31 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: done

File: drake/solvers/Optimization.h (hunk 32 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: yes.  good call.

**Comment by RussTedrake** [comment]
@avalenzu -- don't have a great justification, except that there is a lot of coupling between constraint and optimizationproblem (just tried pulling it out quickly, and decided it required so many forward declarations that it didn't make sense at the moment).  cleaning this up would be good to do, soon. 
one specific thing we have talked about doing is to separate the constraint implement from the decision variable association, with a constraint wrapper class.

File: drake/core/Gradient.h (hunk 4 above)
Comment (tkoolen) [inline drake/core/Gradient.h]: I'm saying I like this better than the cast and that we should use a modified version of this throughout, the modification being support for arbitrarily sized matrices.

File: drake/core/Vector.h (hunk 6 above)
Comment (tkoolen) [inline drake/core/Vector.h]: agree.

File: drake/core/Gradient.h (hunk 4 above)
Comment (RussTedrake) [inline drake/core/Gradient.h]: Got it.  I think that you and I should try to make Gradient.h really good... mostly by moving over the most relevant pieces you have in drakeGradientUtil.h (i started this file by moving the entire gradientUtil over, but I think there is a lot of vestigial gradient logic in there, so thought instead that we can move things over to core as we clean and purge).

File: drake/solvers/CMakeLists.txt (hunk 10 above)
Comment (tkoolen) [inline drake/solvers/CMakeLists.txt]: Ah great. Does that mean we should get rid of gfortran as an ubuntu prereq in the snopt install_prereqs.sh?

Comment (RussTedrake) [inline drake/solvers/CMakeLists.txt]: Yes.  definitely.  good call.

File: drake/solvers/Optimization.h (hunk 15 above)
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Yep, I confused myself. This is fine. Just made sure using this little test program:
```
#include <iostream>

#define EIGEN_INITIALIZE_MATRICES_BY_NAN
#include <Eigen/Dense>
#include <unsupported/Eigen/AutoDiff>

using namespace Eigen;
using namespace std;

int main () {
  Matrix<double, 1, 1> mat1;
  mat1(0) = 2.0;
  double mat2_val = 3.0;
  VectorXd mat2_derivs(3);
  mat2_derivs << 1.0, 2.0, 3.0;

  // dynamic size
  {
    auto mat1_autodiff_dynamic = mat1.cast<AutoDiffScalar<VectorXd>>();
    cout << "mat1_autodiff_dynamic derivative: " << mat1_autodiff_dynamic(0).derivatives().transpose() << endl;
    Matrix<AutoDiffScalar<VectorXd>, 1, 1> mat2_autodiff_dynamic;
    mat2_autodiff_dynamic(0) = AutoDiffScalar<VectorXd>(mat2_val, mat2_derivs);
    auto mat3_autodiff_dynamic = (mat1_autodiff_dynamic * mat2_autodiff_dynamic).eval();
    cout << "mat3_autodiff_dynamic derivatives: " << mat3_autodiff_dynamic(0).derivatives().transpose() << endl;
  }

  // fixed size
  {
    auto mat1_autodiff_fixed = mat1.cast<AutoDiffScalar<Vector3d>>().eval();
    cout << "mat1_autodiff_fixed derivatives: " << mat1_autodiff_fixed(0).derivatives().transpose() << endl;
    Matrix<AutoDiffScalar<Vector3d>, 1, 1> mat2_autodiff_fixed;
    mat2_autodiff_fixed(0) = AutoDiffScalar<VectorXd>(mat2_val, Vector3d(mat2_derivs));
    auto mat3_autodiff_fixed = (mat1_autodiff_fixed * mat2_autodiff_fixed).eval();
    cout << "mat3_autodiff_fixed derivatives: " << mat3_autodiff_fixed(0).derivatives().transpose() << endl;
  }

  return 0;
}
```

Output:
```
mat1_autodiff_dynamic derivative: 
mat3_autodiff_dynamic derivatives: 2 4 6
mat1_autodiff_fixed derivatives: 0 0 0
mat3_autodiff_fixed derivatives: 2 4 6
```

Using the debugger, I verified that the cast call uses this AutoDiffScalar constructor:
```
    /** Conversion from a scalar constant to an active scalar.
      * The derivatives are set to zero. */
    /*explicit*/ AutoDiffScalar(const Real& value)
      : m_value(value)
    {
      if(m_derivatives.size()>0)
        m_derivatives.setZero();
    }
```

So yeah, it sets the derivatives to zero.

Comment (tkoolen) [inline drake/solvers/Optimization.h]: btw, accessing that codetree link requires me to start a 14 day trial of their product.

File: drake/core/Gradient.h (hunk 4 above)
Comment (tkoolen) [inline drake/core/Gradient.h]: Yeah, sounds great. I actually started doing some cool stuff related to that in a branch. Coming soon!

File: drake/solvers/Optimization.h (hunk 15 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: Where did you get a codetree link?  ( I can’t find this comment anywhere on the pull request )?
